rfcs/0013-module-versioning.md

```diff
@@ -0,0 +1,154 @@
+writers for those representation do not mention the type or
+a version. So non-OCaml consumers of these representations
+cannot easily detect the type distinctions in OCaml that gave
+rise to the representations.
+
+When a new version of a type is created, it can be worthwhile
+to retain the older version and associated code. That allows
+producing and consuming representations for older versions of the software.
+
+Suppose we create a module with a name and a version:
+
+```ocaml
+module Some_data = struct
+
+  let name = "some_data"
+
+  module Stable = struct
+
+    module V1 = struct
+
+      let version = 1
+
+      module T = struct
+        type t = ... [@@deriving bin_io,sexp,yojson]
+      end
+
+      module Versioned = struct
+
+        type t =
+          { name : string
+          ; version : int
+          ; t : T.t
+          } [@@deriving bin_io,sexp,yojson]
+
+        let create t =
+          { name
+          ; version
+          ; t
+          }
+
+        let to_yojson (t : T.t) =
+          yojson_of_t @@ create t
+
+        (* similarly for binio, sexp *)
+
+      end
+    end
+  end
+end
+```
+
+Putting the name and version in the type `Versioned.t` means they'll be
+in the `sexp` and `yojson` representations, allowing any software to
+detect type version changes.
+
+Note that the definition of `Versioned` is boilerplate, and could be
+generated with an OCaml ppx.
+
+A new version of the module `V2` consists of a new version number,
+a new submodule `T`, and the same boilerplate. A ppx could
+take the version number and submodule to produce the complete
+module definition. Perhaps it would also take the list of
+serializations needed (often just one or two are needed).
+
+### Serialization restricted to Stable, versioned modules
+
+Serialization should be allowed only for modules following
+this discipline, including the use of the `Stable` submodule
+containing versions.
+
+An exception can be carved out for using `sexp` serialization of data
+from modules without `Stable` and versioning. That kind of
+serialization is useful for logging and other printing from within
+Coda, and is less likely to be used with other software.
+
+### Embracing this discipline
+
+As of this writing, there are over 600 type definitions with `deriving
+bin_io` in the Coda codebase, even more with `sexp`. Only about 60
+type definitions are versioned using the `Stable.Vn` naming discipline
+(and without the explicit version information suggested here). To
+embrace this discipline fully will take some effort.
+
+We can make awareness of the existing recommendation, or the extended
+discipline suggested here a checklist item for Github pull requests.
+
+A ppx to implement the extended discipline will make it easier to comply
+with the standard.
+
+### Syntax for ppx
+
+A use of the ppx might look like:
+
+```ocaml
+[@@versioned_module
+  (`Version 1
+  ,`Type int * string,
+  ,`Serializations [bin_io;sexp;yojson]
+  )
+]
+```
```

Comment:
How will this support other deriving bits? `eq` and `compare` for example.

Reply:
but I like this idea to reduce the boilerplate

Reply:
I think we can add as many as we like, but only the serializations will trigger additional code.